pow/src/main/java/tech/pegasys/teku/pow/Web3jEth1Provider.java

```diff
@@ -162,4 +190,46 @@ public Web3jEth1Provider(final Web3j web3j, final AsyncRunner asyncRunner) {
               return (List<EthLog.LogResult<?>>) (List) logs;
             });
   }
+
+  @Override
+  public SafeFuture<Boolean> validate() {
```

Comment:
Spent a bit of time working out how to pull all the advice below together and this is what I come up with to replace this whole method:

```
  @Override
  public SafeFuture<Boolean> validate() {
    if (validating.compareAndSet(false, true)) {
      LOG.debug("Validating endpoint {} ...", this.id);
      return validateChainId()
          .thenCompose(
              result -> {
                if (result == Result.failed) {
                  return SafeFuture.completedFuture(result);
                } else {
                  return validateSyncing();
                }
              })
          .thenApply(
              result -> {
                updateLastValidation(result);
                return result == Result.success;
              })
          .exceptionally(
              error -> {
                LOG.warn(
                    "Endpoint {} is INVALID | {}",
                    this.id,
                    Throwables.getRootCause(error).getMessage());
                updateLastValidation(Result.failed);
                return false;
              });
    } else {
      LOG.debug("Already validating");
      return SafeFuture.completedFuture(isValid());
    }
  }

  private SafeFuture<Result> validateChainId() {
    return getChainId()
        .thenApply(
            chainId -> {
              if (chainId.intValueExact() != Constants.DEPOSIT_CHAIN_ID) {
                STATUS_LOG.eth1DepositChainIdMismatch(
                    Constants.DEPOSIT_CHAIN_ID, chainId.intValueExact(), this.id);
                return Result.failed;
              }
              return Result.success;
            });
  }

  private SafeFuture<Result> validateSyncing() {
    return ethSyncing()
        .thenApply(
            syncing -> {
              if (syncing) {
                LOG.warn("Endpoint {} is INVALID | Still syncing", this.id);
                updateLastValidation(Result.failed);
                return Result.failed;
              } else {
                LOG.debug("Endpoint {} is VALID", this.id);
                updateLastValidation(Result.success);
                return Result.success;
              }
            });
  }
```

Reply:
good point on splitting things up..